adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +59,145 @@ def as_dict(self: Self) -> dict: ...
 
 
 class DefaultRequester(Requester):
-    def __init__(self: Self) -> None: ...
+    __slots__ = ("_credentials", "api_root", "client", "retries", "retry_interval")
+
+    def __init__(
+        self: Self,
+        base_url: str,
+        root_path: str = "/api/v2/",
+        timeout: float = 5.0,
+        retries: int = 5,
+        retry_interval: float = 5.0,
+    ) -> None:
+        self.retries = retries
+        self.retry_interval = retry_interval
+        self.api_root = self._make_url(root_path, base=base_url)
+        self.client = httpx.AsyncClient(timeout=timeout)
+
+    async def login(self: Self, credentials: Credentials) -> Self:
+        login_url = self._make_url("login", base=self.api_root)
+
+        try:
+            response = await self._do_request(self.client.post(url=login_url, data=credentials.dict()))
+        except UnauthorizedError as e:
+            raise WrongCredentialsError from e
+
+        if response.status_code != 200:
+            message = f"Authentication error: {response.status_code} for url: {login_url}"
+            raise LoginError(message)
+
+        self._credentials = credentials
+        return self
+
+    async def get(self: Self, *path: str | int, query_params: dict | None = None) -> HTTPXRequesterResponse:
+        return await self.request(*path, method=self.client.get, params=query_params or {})
+
+    async def post(self: Self, *path: str | int, data: dict) -> HTTPXRequesterResponse:
+        return await self.request(*path, method=self.client.post, data=data)
+
+    async def patch(self: Self, *path: str | int, data: dict) -> HTTPXRequesterResponse:
+        return await self.request(*path, method=self.client.patch, data=data)
+
+    async def delete(self: Self, *path: str | int) -> HTTPXRequesterResponse:
+        return await self.request(*path, method=self.client.delete)
+
+    async def request(self: Self, *path: str | int, method: Callable, **kwargs: dict) -> HTTPXRequesterResponse:
+        url = self._make_url(*path, base=self.api_root)
+
+        try:
+            response = await self._do_request(method(url, **kwargs))
+        except UnauthorizedError:
+            await self._reconnect()
+            response = await self._do_request(method(url, **kwargs))
+
+        return HTTPXRequesterResponse(response=response)
+
+    @staticmethod
+    def _make_url(*path: str | int, base: str) -> str:
+        return urljoin(base, "/".join(map(str, (*path, ""))))
+
+    @convert_exceptions
+    async def _do_request(self: Self, request_coro: Coroutine[Any, Any, httpx.Response]) -> httpx.Response:
+        return await request_coro
+
+    async def _reconnect(self: Self) -> None:
+        credentials = self._ensure_credentials()
+
+        for _ in range(self.retries):
+            try:
+                await self.login(credentials=credentials)
+                break
+            except (httpx.NetworkError, httpx.TransportError):
+                await sleep(delay=self.retry_interval)
```

Comment:
🤓 
if it was last attempt, there's no meed to sleep

Reply:
fixed